sklearn/linear_model/sag_fast.pyx

```diff
@@ -0,0 +1,325 @@
+
+                if fit_intercept:
+                    intercept_sum_gradient += \
+                        gradient - gradient_memory[current_index]
+                    intercept -= (eta *
+                                  (intercept_sum_gradient / num_seen) *
+                                  intercept_decay)
+
+                # update the gradient memory for this sample
+                gradient_memory[current_index] = gradient
+
+                wscale *= one_minus_eta_alpha
+
+                if itr == 0:
+                    cumulative_sums[0] = eta / (wscale * num_seen)
+                else:
+                    cumulative_sums[itr] = (cumulative_sums[itr - 1]
+                                            + eta / (wscale * num_seen))
+
+                # if wscale gets too small, we need to reset the scale
+                if wscale < 1e-9:
+                    if verbose:
+                        with gil:
+                            print("rescaling...")
+                    scale_weights(weights, wscale, n_features, n_samples,
+                                  itr, cumulative_sums, feature_hist,
+                                  sum_gradient)
+                    wscale = 1.0
+
+                total_iter += 1
+
+            # check if the stopping criteria is reached
+            scale_weights(weights, wscale, n_features, n_samples, n_samples - 1,
+                          cumulative_sums, feature_hist, sum_gradient)
+            wscale = 1.0
+            
+            max_change = 0.0
+            max_weight = 0.0
+            for j in range(n_features):
+                max_weight = fmax(max_weight, fabs(weights[j]))
+            for j in range(n_features):
+                max_change = fmax(max_change,
+                                  fabs(weights[j] - previous_weights[j]))
+                previous_weights[j] = weights[j]
+
+            if max_change / max_weight <= tol:
+                if verbose:
+                    end_time = time(NULL)
+                    with gil:
+                        print("convergence after %d epochs took %d seconds" %
+                              ((total_iter / n_samples),
+                              (end_time - start_time)))
+                break
+
+            if total_iter / n_samples >= max_iter:
+                if verbose:
+                    end_time = time(NULL)
+                    with gil:
+                        print(("max_iter reached after %d seconds") %
+                              (end_time - start_time))
+                max_iter_reached = True
+                break
+
+
+    if infinity:
+        raise ValueError(("Floating-point under-/overflow occurred at epoch"
+                          " #%d. Lowering the eta0 or scaling the input data"
+                          " with StandardScaler or"
+                          " MinMaxScaler might help.") % (total_iter + 1))
+
+    return intercept, num_seen, max_iter_reached, intercept_sum_gradient
```

Comment:
Actually it would even be better to have a variable `n_iter = 0` initialized at the beginning instead of `total_iter / n_updates` and then increment it by one after each epoch (instead of each update). The we never need to divide it by `n_samples`.